packages/xod-client/src/project/selectors.js

```diff
@@ -71,10 +71,22 @@ export const getCurrentPatch = createSelector(
 );
 
 export const getDeducedPinTypes = createMemoizedSelector(
-  [getProject, getCurrentPatch],
-  [R.equals, R.equals],
-  (project, maybeCurrentPatch) =>
-    foldMaybe({}, patch => XP.deducePinTypes(patch, project), maybeCurrentPatch)
+  [getProject, getCurrentPatchPath],
+  [
+    (prev, next) => {
+      const prevPatches = XP.listPatches(prev);
+      const nextPatches = XP.listPatches(next);
+      return XP.patchListEqualsBy(XP.samePatchValidity)(
+        prevPatches,
+        nextPatches
+      );
+    },
+  ],
```

Comment:
Inlining the comparators makes `getDeducedPinTypes` harder to follow. Suggest extracting to a function